### ISSUE TITLE
[Suggestion] Have TrainLastCheckpoint in addition/instead of TrainEndCheckpoint

### ISSUE THREAD

**Issue by YannDubs** [skorch-dev/skorch#529]
Wouldn't it make more sense to have a `TrainLastCheckpoint` rather than `TrainEndCheckpoint`? I.e.use `on_epoch_end` instead of `on_train_end`. Indeed if you kill the job early, it would be nice to have the last model checkpointed to be able to start again from it. Does checkpointing take a long time ?

**Comment by BenjaminBossan** [comment]
Good observation.

> Indeed if you kill the job early, it would be nice to have the last model checkpointed to be able to start again from it.

We thought about having a special callback method that is called on exception. This way, we could  change `TrainEndCheckpoint` to also create a checkpoint on `KeyBoardInterrupt`. What are your thoughts on that?

> Does checkpointing take a long time ?

This can probably not be answered in general. If training one epoch takes very long but the model is small, the time is probably negligible. If the model is large but training one epoch is short, it could introduce significant overhead. Who knows, it might even depend on your disk. I wouldn't change the existing callback to save after each epoch, instead I would either make it optional or create a separate callback.

Since I haven't worked much with checkpoints, maybe @thomasjpfan or @ottonemo could add their thoughts on this.

**Comment by YannDubs** [comment]
Catching the exception is probably the cleanest way of doing. But I don't think that `KeyBoardInterrupt` is sufficient. I think many people submit jobs via workload managers such as SLURM, in which case it would be good to save the last model when encountering a timeout or pre-emption. Maybe the most general way (but not the cleanest) would be to catch any exception and save a checkpoint before exiting.

**Comment by thomasjpfan** [comment]
What do you think of catching any exception here:

https://github.com/skorch-dev/skorch/blob/d4865aa7868e3c9b73fc6be699303db4560c6f2f/skorch/net.py#L799-L802

which will allow `on_train_end` to trigger, allowing `TrainEndCheckpoint` to checkpoint?

**Comment by ottonemo** [comment]
I think that job schedulers like SLURM are a good argument *against* handling the exception and opting for a callback that saves the model at every epoch instead. For example, SLURM implements preemption by sending `SIGTERM` to the process which can obviously be caught but we would need to introduce signal handling logic to the training loop. It probably doesn't stop there since there are platforms that might do signalling differently (not sure about Windows, for example).

It is unlikely that someone trains a model that takes long to store with very little data. What is far more likely is that the checkpoints are stored on network filesystems (or buckets) which might need more time to store the data. In this case the user would introduce a significant overhead when enabling a `TrainLastCheckpoint` callback if the data is small. Since the store operation is IO-bound and the rest of the process is not, we could use async IO to write a model copy in the background. Of course this has its own problems (e.g., writing to the same file in parallel) which would need to be addressed but the point here is that there is a workable solution if people start to complain about performance issues with this type of callback.

So in essence: I'm in favor of having a `TrainLastCheckpoint` callback.

**Comment by thomasjpfan** [comment]
The checkpoint has a supports a `monitor=None` which will checkpoint at the end of every epoch.

**Comment by YannDubs** [comment]
I agree with @ottonemo , saving at every epoch might be the most general way and platform independent. Thanks @thomasjpfan, I hadn't realized! Then I'm happy to close this.

**Comment by YannDubs** [comment]
Actually, @thomasjpfan if you use `monitor=None` will it automatically load the last checkpoint if you use `LoadInitState` ?

BTW I still think `TrainLastCheckpoint` would be nice as it will only keep the last checkpoint saved, making the directory cleaner.

**Comment by BenjaminBossan** [comment]
> Since the store operation is IO-bound and the rest of the process is not, we could use async IO to write a model copy in the background.

Just for my understanding: When this happens asynchronously but the underlying parameters are updated by the training process while persisting, what would happen? Would we need to make a copy of the parameters first?

**Comment by ottonemo** [comment]
> > Since the store operation is IO-bound and the rest of the process is not, we could use async IO to write a model copy in the background.
> 
> Just for my understanding: When this happens asynchronously but the underlying parameters are updated by the training process while persisting, what would happen? Would we need to make a copy of the parameters first?

Yes, that is one of the necessary mitigations to avoid defective model checkpoints. These copies shouldn't be a too hard hit on performance though, since they are (in worst-case) moving data from compute device to main memory which is still significantly faster than disk or network IO. Locking the checkpoint file access is the second part of such an asynchronous checkpointing solution.

**Comment by BenjaminBossan** [comment]
> Actually, @thomasjpfan if you use `monitor=None` will it automatically load the last checkpoint if you use `LoadInitState` ?
> BTW I still think TrainLastCheckpoint would be nice as it will only keep the last checkpoint saved, making the directory cleaner.

By default, it should overwrite the latest checkpoint, so it would automatically load the last checkpoint. Do you still see the need for a `TrainLastCheckpoint`?

> Yes, that is one of the necessary mitigations to avoid defective model checkpoints. These copies shouldn't be a too hard hit on performance though, since they are (in worst-case) moving data from compute device to main memory which is still significantly faster than disk or network IO. Locking the checkpoint file access is the second part of such an asynchronous checkpointing solution.

This sure sounds like a non-trivial problem. Since I haven't worked with asyncio, someone else would need to step up to get this to work correctly.

**Comment by YannDubs** [comment]
You are right ! I just checked and it does overwrite the latest checkpoint. Then I guess `TrainLastCheckpoint ` is not needed anymore. I will close the issue.

As last comment regarding the `monitor=None`, currently if you want to keep a checkpoint for every epochs you have to manually change `f_params `, `f_optimizer`, `f_history ` using `{last_epoch}`. I think that it would be better to also let `fn_prefix` use `net`, `last_epoch`, or `last_batch` such that you can modify the 3 at once using `fn_prefix`. It's not crucial but would be a little better.

**Comment by BenjaminBossan** [comment]
> A last comment regarding the `monitor=None`: currently if you want to keep a checkpoint for every epochs you have to manually change `f_params `, `f_optimizer`, `f_pickle` using `{last_epoch}`. I think that it would be better to also let `fn_prefix` use `net`, `last_epoch`, or `last_batch` such that you can modify the 3 at once using `fn_prefix`. It's not crucial but would be a little better.

That sounds like a good idea to me, would you want to add that feature?
